add Nordic JEDEC QSPI NOR flash driver

kamlaz opened the pull request:
Add QSPI flash driver support for Nordic nRF52840 chip (PCA10056 development board).

pabigot committed:
samples: drivers: spi_flash: switch nrf52840_pca10056 to Nordic QSPI

Adjust the configuration file, disable the SPI
driver and enable the QSPI driver and flash node.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>
Signed-off-by: Kamil Lazowski <Kamil.Lazowski@nordicsemi.no>

kamlaz committed:
drivers: flash: add Nordic JEDEC QSPI NOR flash driver

Most JEDEC NOR flash devices uses not only typical SPI mode
(MISO,MOSI,SCK and CS), but also QSPI mode (IO0,IO1,IO2,IO3,SCK and CS).
QSPI mode uses more data lines and as a result provide higher
throughput. If this were not enough, Nordic chips provide
hardware acceleration for read/write/erase functions, what
gives significant performance boost.
It does a lot of things "behind the scene", i.e when user has written
some data to the flash and would like to read them back, it has to wait
until the flash is ready by reading WIP bit in Status Register.
This driver does it automatically.

Signed-off-by: Kamil Lazowski <Kamil.Lazowski@nordicsemi.no>

ulfalizer commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,37 @@
+	select FLASH_HAS_DRIVER_ENABLED
+	depends on HAS_HW_NRF_QSPI
+
+if QSPI_NOR
```
Add a blank line after the `if` to make it balanced with the `endif`. See https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#header-comments-and-other-nits.

ulfalizer commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,37 @@
+	help
+	  Enable support for nrfx QSPI driver with EasyDMA.
+
+config	QE_BIT
```
Accidental tab instead of space.

ulfalizer commented on dts/bindings/flash_controller/nordic,nrf-qspi.yaml:
```diff
@@ -0,0 +1,39 @@
+    type: array
+    required: true
+    description: |
+      Pin numbers associated with chip-select signals
```
Missing newline.

ulfalizer commented on dts/bindings/mtd/nordic,qspi-nor.yaml:
```diff
@@ -0,0 +1,85 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Any QSPI NOR flash that supports the JEDEC CFI interface.
```
This will show up in the header as a comment above the generated macros, so might be better to make it "QSPI NOR flash supporting the JEDEC CFI interface".

ulfalizer reviewed (commented):
Some nits.

nvlsianpu commented on samples/drivers/spi_flash/boards/nrf52840_pca10056.conf:
```diff
@@ -1,7 +1,12 @@
 #
-# Copyright (c) 2019 Peter Bigot Consulting, LLC
```
rather add copyright, not replace

pabigot replied:
This commit is actually from me, but authorship got lost during development, and I removed  this because in the churn around this I'd forgotten it was originally added on my own time not for Nordic.  But yes, it probably should have both, just for history.

I would rather that the commit was restored to the original authorship and sign-off, if that can be done.  `git commit --amend -C sha-of-original` should do it.

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+};
+
+/**
+ * struct qspi_nor_data - Structure for defining the QSPI NOR access
```
@brief, let's keep doxygen comment consistent

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+	struct k_sem sync;
+	bool write_protection;
+};
+
```
to many blank lines

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+
+static inline int qspi_get_lines_read(u8_t lines)
+{
+	register int ret = -EINVAL;
```
minor suggestion: LUT maybe?
```
const int ret_array[] = {NRF_QSPI_READOC_FASTREAD, NRF_QSPI_READOC_READ2O, NRF_QSPI_READOC_READ2IO, NRF_QSPI_READOC_READ4O, NRF_QSPI_READOC_READ4IO};

if (lines < ARRAY_SIZE( ret_array[])) {
  ret = ret_array[lines];
}
```

above as well

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+/**
+ * @brief QSPI handler
+ *
+ * @param event - driver event type
```
remove dash.
in other places of this file is saw colon after a perimeter line, I prefer consistency like `@param param_name what_param_does`

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+/**
+ * @brief Fills init struct
+ *
+ * @param config		- Pointer to the config struct provided by user
```
to many white space

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+
+static const struct qspi_nor_config flash_id = {
+	.id = DT_INST_0_NORDIC_QSPI_NOR_JEDEC_ID,
+	.size = DT_INST_0_NORDIC_QSPI_NOR_SIZE / 8,
```
```suggestion
	.size = INST_0_BYTES
```

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+
+/* instance 0 size in bytes */
+#define INST_0_BYTES (DT_INST_0_NORDIC_QSPI_NOR_SIZE / 8)
```
move INST_0_BYTES outside of ifdefry in case using in device initialization macro

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,690 @@
+ */
+static inline nrf_qspi_frequency_t get_nrf_qspi_prescaler(u32_t frequency)
+{
+	if (frequency < 2130000) {
```
Does it make sense to use LUT in the loop searching here?

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,688 @@
+			   QSPI_IS_BLOCK_ALIGNED(addr)) {
+			/* 64 kB block erase */
+			qspi_lock(dev);
+			res = nrfx_qspi_erase(NRF_QSPI_ERASE_LEN_64KB, addr);
```
`res` codes for partial erases are not checked in this loop, only the translated code for the last result is returned.

kamlaz replied:
That is right - I missed that.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,688 @@
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	while (size) {
+		if (size == params->size) {
```
`addr` is not checked here, also for sizes greater then the size of the chip the loop will try to do partial erases.

kamlaz replied:
I have added check condition.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,688 @@
+
+	qspi_fill_init_struct(&QSPIconfig);
+
+	int res = nrfx_qspi_init(&QSPIconfig, qspi_handler, dev);
```
In case this is not successful, the function should no attempt to call `qspi_send_cmd`.

kamlaz replied:
Fixed

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,688 @@
+		cmd.rx_buf = NULL;
+
+		if (qspi_send_cmd(dev, &cmd) != 0) {
+			res = -EIO;
```
This code will get translated in `qspi_get_zephyr_ret_code` to `-EBUSY`. I guess this was not the intent.

kamlaz replied:
Exactly - silly mistake.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,688 @@
+				   const struct qspi_nor_config *const flash_id)
+{
+	u8_t rx_b[QSPI_NOR_MAX_ID_LEN];
+	struct qspi_buf q_rx_buf = {
```
This can be `const`.

kamlaz replied:
Added `const`

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,688 @@
+		.buf = rx_b,
+		.len = QSPI_NOR_MAX_ID_LEN
+	};
+	struct qspi_cmd cmd = {
```
This can be `const`.

kamlaz replied:
Added `const`

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,688 @@
+		.len = size
+	};
+
+	int ret = nrfx_qspi_read(rx_buf.buf, rx_buf.len, addr);
```
What's the point in involving `rx_buf` while you could use `dest` and `size` directly here?
Same goes for the call to `nrfx_qspi_write` below.

kamlaz replied:
Indeed, it is not needed anymore. Removed it.

pabigot commented on samples/drivers/spi_flash/boards/nrf52840_pca10056.conf:
```diff
@@ -1,7 +1,12 @@
+# Disable the SPI NOR driver, then switch
+# to QSPI NOR driver.
+CONFIG_SPI_2=n
+CONFIG_SPI_NOR=n
```
The board-specific overlay is appended to the project overlay which selects `CONFIG_SPI_NOR`, so the overlay needs to deselect it.

If `CONFIG_SPI_2` remains enabled I believe something else broke, but I don't remember what it was.  Probably to do with the subordinate node on that device, or the flash partitions bound to it.

anangl replied:
I don't think I understand. Which project overlay are you referring to? This sample's [prj.conf](https://github.com/zephyrproject-rtos/zephyr/blob/f638a7e637a341a7f3e25ce281cfa6957ab3d879/samples/drivers/spi_flash/prj.conf) does not select `CONFIG_QSPI_NOR`.

pabigot replied:
OK, I see; this is a result of making the overlay consistent with the other samples that are spi-nor specific.  

I'd say try removing the two lines and see what happens.

kamlaz replied:
Removing `CONFIG_SPI_NOR` has no effect, but removing `CONFIG_SPI_2` causes that application is unable to find driver using `device_get_binding`.  Perhaps it is due to pins assigned to that `SPI` instance (CS pins) are also used by QSPI.

anangl replied:
So the problem is invalid default configuration of the board (same pins assigned to `spi2` and `qspi`, and both these peripherals enabled). And that is what should be corrected. By disabling `SPI_2` in this application you are only masking the problem.

kamlaz commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,38 @@
+# Copyright (c) 2019, Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig QSPI_NOR
```
Good idea. I will change it.

kamlaz commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,38 @@
+	help
+	  Enable support for nrfx QSPI driver with EasyDMA.
+
+config QE_BIT
```
Yes, it should be changed.

kamlaz commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,38 @@
+
+if QSPI_NOR
+
+config QSPI_NRF_QSPI
```
Right - it could be removed.

anangl commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,34 @@
+if NORDIC_QSPI_NOR
+
+config QSPI_NOR_QE_BIT
+	int "Quad Enable bit number in Status Regiter"
```
typo - Register

kamlaz replied:
Fixed

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,687 @@
+				size -= QSPI_SECTOR_SIZE;
+			} else {
+				/* minimal erase size is at least a sector size */
+				LOG_DBG("unsupported at 0x%lx size %zu", (long)addr, size);
```
Errors should be rather reported with `LOG_ERR`.

kamlaz replied:
Changed `LOG_DBG` to `LOG_ERR`.

kamlaz commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,34 @@
+
+if NORDIC_QSPI_NOR
+
+config QSPI_NOR_QE_BIT
```
Changed prefixes.

kamlaz commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,687 @@
+#define QSPI_NOR_MAX_ID_LEN                     SPI_NOR_MAX_ID_LEN
+
+/* Status register bits */
+#define QSPI_NOR_QE_BIT                         BIT(CONFIG_QSPI_NOR_QE_BIT)
```
Changed.

kamlaz commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,687 @@
+ */
+static inline nrf_qspi_frequency_t get_nrf_qspi_prescaler(u32_t frequency)
+{
+	if (frequency < 2130000) {
```
Fixed.

kamlaz commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,687 @@
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((addr + size) > params->size)) {
+		return -EINVAL;
```
Indeed. Fixed.

kamlaz commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,687 @@
+}
+
+/* Configures QSPI memory for the transfer */
+int qspi_nrfx_configure(struct device *dev)
```
Added `static`

kamlaz commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,687 @@
+
+	/* now the spi bus is configured, we can verify the flash id */
+	if (qspi_nor_read_id(dev, params) != 0) {
+		return -ENODEV;
```
I have added `LOG_ERR` inside the `qspi_nor_read_id` to provide info about extected ID an obtained ID.

kamlaz commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,687 @@
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((addr + size) > params->size)) {
```
Added same address check algorithm as in soc_flash_nrf.c

anangl replied:
It seems that you did, but only in `qspi_erase`. Was that the intent?

nvlsianpu replied:
@kamlaz please apply hat was requested here - it's the bug

kamlaz replied:
I changed that also in write and read commands. I just missed them before.

carlescufi commented:
@anangl @pabigot please re-review.

nvlsianpu commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,673 @@
+{
+	register int ret = -EINVAL;
+
+	if (frequency > 2000000) {
```
this comparison stages look erroneous as ensure else case in second compassion.

nvlsianpu replied:
just 2000000ul

pabigot replied:
IMO this should be `2000000UL`, i.e. upper case.  We don't want a lower case `l` because it looks too much like `1`, so for consistency it should be `UL`.  Also see line 172 where `32000000uL` looks weird.

Not a blocking comment, just a preference.

kamlaz replied:
I have added changes due to yours request.

pabigot reviewed (commented):
Just one comment for now.  I'll approve when I rebase my littlefs test support on this and verify the behavior again.

pabigot committed:
samples: littlefs: correct to use SPI NOR flash memory on nrf52840_pca10056

A configuration file attempted to select the external flash memory for
this platform, but there was no overlay that redefined the storage
partition to be on that device rather than the SOC flash.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot reviewed (changes requested):
While testing I noticed that `samples/subys/fs/littlefs` for nrf52840_pca10056 has a configuration override that targets the external flash memory, but did not include a devicetree overlay that redefined the partition table so the storage partition is on that device.

Please look at https://github.com/pabigot/zephyr/commits/pr/21535 and cherry-pick https://github.com/pabigot/zephyr/commit/f867c8782354cb48d50f08785a2c5b641b12882a into your PR before the other two commits.  Then your changes to switch from SPI to QSPI for this sample will actually do something.

Alternatively I can submit that fix separately, we can wait until it's merged, then merge this PR.

carlescufi commented:
@nvlsianpu please re-review

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,673 @@
+	}
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((size + addr) > params->size)) {
```
This is also checked in `qspi_erase` in a bit more extensive way, why do it twice?
And `-EINVAL` is returned there, as well as in `qspi_nor_read` and `qspi_nor_write`, which seems more suitable.

kamlaz replied:
Right. I reduced this check to one and moved it to the `qspi_nor_erase`.

anangl commented on drivers/flash/nrf_qspi_nor.c:
```diff
@@ -0,0 +1,673 @@
+
+	if (frequency < 2000000UL) {
+		ret = -EINVAL;
+	} else if (frequency > 32000000UL) {
```
It can be `>=`.

kamlaz replied:
Fixed

nvlsianpu reviewed (commented):
Possible overflows of `addr + size` is not handled properly for `write` operation.

anangl commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,34 @@
+	help
+	  Device driver initialization priority.
+	  Device is connected to SPI bus, it has to
+	  be initialized after SPI driver.
```
This is not true, this device does not use an SPI driver.

kamlaz replied:
I removed misleading part.

anangl commented on drivers/flash/Kconfig.nordic_qspi_nor:
```diff
@@ -0,0 +1,34 @@
+	default 6
+
+config NORDIC_QSPI_NOR_INIT_PRIORITY
+	int
```
As this option has no prompt, users will not be able to modify it in menuconfig or in an application's `prj.conf`. Was that the intent? I see that in `Kconfig.nor` the situation is the same, but I'm actually not sure if that was intended there either.

kamlaz replied:
I did it intentionally to provide consistency with `Kconfig.nor`. It could be changed, but I think both changes (in `Kconfig.nordic_qspi_nor` and `Kconfig.nor`) schould be done in one commit in another PR (since this one does not touch `spi_nor`).

pabigot replied:
I agree with that.  I doubt it was intentional in either case.  I can change it in both later, or we could make progress on https://github.com/zephyrproject-rtos/zephyr/issues/20348#issuecomment-564108715 which could lead to automatic management of initialization order based on dependencies.
